File: typescript/packages/core/tests/failure.test.ts
import { expect, test } from 'vitest'
import { CellMismatchError, ReturnShapeError } from '../src/cell-diff.ts'
import { compareDocString } from '../src/doc-string-diff.ts'
import { toFailure } from '../src/failure.ts'
import { spanFromOffsets } from '../src/span.ts'

test('toFailure extracts cells from a CellMismatchError', () => {
  const source = 'a | 5 |'
  const err = new CellMismatchError([
    { column: 'n', span: spanFromOffsets(source, 4, 5), expected: '5', actual: '4', ok: false },
  ])
  const f = toFailure(err, 'oath.md', 3)
  expect(f.cells).toEqual([{ from: 4, to: 5, actual: '4' }])
  expect(typeof f.message).toBe('string')
  expect(typeof f.stack).toBe('string')
})

test('toFailure extracts a doc-string mismatch as a cell', () => {
  const source = 'Hello!\n'
  const diff = compareDocString('Goodbye!\n', 'Hello!\n', spanFromOffsets(source, 0, 7))
  const f = toFailure(new CellMismatchError([diff!]), 'oath.md', 3)
  expect(f.cells).toEqual([{ from: 0, to: 7, actual: '"Goodbye!\\n"' }])
})

test('toFailure leaves cells undefined for a plain error or ReturnShapeError', () => {
  expect(toFailure(new Error('nope'), 'oath.md', 3).cells).toBeUndefined()
  expect(toFailure(new ReturnShapeError('bad'), 'oath.md', 3).cells).toBeUndefined()
})

test('toFailure reads the failing line from an injected stack frame, else falls back', () => {
  const err = new Error('boom')
  err.stack = 'Error: boom\n    at handler (steps.ts:1:1)\n    at step (docs/a.md:12:3)'
  expect(toFailure(err, 'docs/a.md', 99).line).toBe(12)

  const noFrame = new Error('boom')
  noFrame.stack = 'Error: boom\n    at handler (steps.ts:1:1)'
  expect(toFailure(noFrame, 'docs/a.md', 99).line).toBe(99)
})

test('toFailure regex-escapes the oath path (a dot is literal)', () => {
  const err = new Error('boom')
  // 'X' stands in for the dot: if the oath path's `.` were treated as a regex
  // wildcard it would match this frame; escaped, it must not.
  err.stack = 'Error: boom\n    at step (aXmd:7:1)'
  // oathPath "a.md" must NOT match "aXmd"
  expect(toFailure(err, 'a.md', 42).line).toBe(42)
})

File: rust/core/tests/failure_test.rs
//! Port of `FailureTest.java` / `failure.test.ts`. Java's synthetic
//! `StackTraceElement` frame becomes a structural [`FailureLocation`]; the
//! regex-escape case becomes an exact path-match check. The "message/stack is a
//! String" type assertions are dropped (type-level in Rust).

use varar_core::cell_diff::CellDiff;
use varar_core::doc_string_diff::compare_doc_string;
use varar_core::error::{FailureLocation, HandlerError, StepError, StepFailure};
use varar_core::failure::to_failure;
use varar_core::result::CellFailure;
use varar_core::span::Span;

fn located(error: StepError, path: &str, line: usize) -> StepFailure {
    StepFailure {
        error,
        location: Some(FailureLocation {
            label: String::new(),
            path: path.to_string(),
            line,
        }),
    }
}

#[test]
fn to_failure_extracts_cells_from_a_cell_mismatch() {
    let source = "a | 5 |";
    let sf = StepFailure::bare(StepError::CellMismatch(vec![CellDiff::new(
        "n",
        Span::from_offsets(source, 4, 5),
        "5",
        "4",
        false,
    )]));
    let f = to_failure(&sf, "oath.md", 3);
    assert_eq!(Some(vec![CellFailure::new(4, 5, "4")]), f.cells);
}

#[test]
fn to_failure_extracts_a_doc_string_mismatch_as_a_cell() {
    let source = "Hello!\n";
    let diff = compare_doc_string(
        Some(&varar_core::value::Value::from("Goodbye!\n")),
        "Hello!\n",
        Span::from_offsets(source, 0, 7),
    )
    .unwrap()
    .unwrap();
    let sf = StepFailure::bare(StepError::CellMismatch(vec![diff]));
    let f = to_failure(&sf, "oath.md", 3);
    assert_eq!(Some(vec![CellFailure::new(0, 7, "\"Goodbye!\\n\"")]), f.cells);
}

#[test]
fn to_failure_leaves_cells_null_for_a_plain_exception_or_return_shape() {
    let plain = StepFailure::bare(StepError::Handler(HandlerError::new("nope")));
    assert_eq!(None, to_failure(&plain, "oath.md", 3).cells);
    let shape = StepFailure::bare(StepError::ReturnShape("bad".to_string()));
    assert_eq!(None, to_failure(&shape, "oath.md", 3).cells);
}

#[test]
fn to_failure_reads_the_failing_line_from_an_injected_location_else_falls_back() {
    let with_frame = located(StepError::Handler(HandlerError::new("boom")), "docs/a.md", 12);
    assert_eq!(12, to_failure(&with_frame, "docs/a.md", 99).line);

    let no_frame = StepFailure::bare(StepError::Handler(HandlerError::new("boom")));
    assert_eq!(99, to_failure(&no_frame, "docs/a.md", 99).line);
}

#[test]
fn to_failure_uses_an_exact_oath_path_match() {
    // 'aXmd' must not be treated as matching oath path 'a.md' (Java escapes the
    // regex dot; Rust compares paths by `==`).
    let sf = located(StepError::Handler(HandlerError::new("boom")), "aXmd", 7);
    assert_eq!(42, to_failure(&sf, "a.md", 42).line);
}
